### ISSUE TITLE
_fc and _att features [clarification question]

### ISSUE THREAD

**Issue by nilinykh** [ruotianluo/self-critical.pytorch#224]
Hej,
thank you for the great work! Much appreciated!
I have a question about `cocobu_fc` / `cocobu_att` and `cocotalk_fc` / `cocotalk_att` pairs: why did you specifically extract fc features here? I mean, we are using the ones ending with `_att` in the model, right? These features are actually the convolutional features. But is there any additional use to the `_fc` features? I was wondering about it, because they are not used anywhere in general + plus when I run `eval.py`, for example, the data that it uses has empty `_fc` features. I guess it does not matter, since we are not using them, but still, why do we even have them mentioned...

**Comment by ruotianluo** [comment]
they are used in fc model and updown model.

**Comment by nilinykh** [comment]
Ah, I see, I think I found where exactly it is in the code, so all good, thanks.

I have another question, more of just learning about image captioning transformer: let's say, I want to visualise attention in a single head of the first layer of the encoder.
So, in each encoder layer's self-attention I can get the softmax scores (right before their multiplication with V), these scores have the shape of `batch_size x num_heads x 196 x 196`. '196' here comes from the fact that I am using top-down with 14x14 grid. My scores could be, for example, of shape `10 x 8 x 196 x 196`.

Now I want to map these scores into my image (visualise this head's attention). However, my image is 14x14, but my attention scores have 1 additional dimension with 196 due to the nature of softmax (10 x 8 x *196 x *196*).
Do you perhaps know what one could do in this case? How can I map these scores into an image, which is of different shape? I am a little confused with how to interpret the softmax scores: are they attention weights over 196 x 196 grid, where each object has a score that shows how much it attend to itself or another object?

**Comment by ruotianluo** [comment]
each grid has an attention map over all the grids.

**Comment by nilinykh** [comment]
you mean to say, that each individual cell in the grid attends to all cells in the grid? this is why we have 196 x 196?

**Comment by ruotianluo** [comment]
I guess what I mean grid is the same as what you mean cell. so yes.
